File: gaia-sdk-python/gaia_sdk/api/rx/rx.py
from typing import Callable, List, TypeVar

from rx.core.typing import Observable

import rx
from gaia_sdk.graphql import QueryRes, MutationRes
from gaia_sdk.graphql.GaiaResponse import QueryResponse, MutationResponse
from rx import operators as ops, pipe

T = TypeVar('T')


def mapQ(observable: Observable[QueryResponse], mapper: Callable[[QueryRes], T]) -> Observable[T]:
    def flat_map(obj: QueryResponse):
        if obj.errors and len(obj.errors) > 0:
            return rx.throw(obj.errors[0]["message"])
        return rx.of(mapper(obj.data))

    return pipe(ops.map(lambda x: QueryResponse(x)),
                ops.flat_map(flat_map))(observable)


def flat_mapQ(observable: Observable[QueryResponse], mapper: Callable[[QueryRes], List[T]]) -> Observable[T]:
    def flat_map(obj: QueryResponse):
        if obj.errors and len(obj.errors) > 0:
            return rx.throw(obj.errors[0]["message"])
        return rx.from_iterable(mapper(obj.data))

    return pipe(ops.map(lambda x: QueryResponse(x)),
                ops.flat_map(flat_map))(observable)


def mapM(observable: Observable[MutationResponse], mapper: Callable[[MutationRes], T]) -> Observable[T]:
    def flat_map(response: MutationResponse):
        if response.errors and len(response.errors) > 0:
            return rx.throw(response.errors[0]["message"])
        return rx.of(mapper(response.data))

    return pipe(ops.map(lambda x: MutationResponse(x)),
                ops.flat_map(flat_map))(observable)


def flat_mapM(observable: Observable[MutationResponse], mapper: Callable[[MutationRes], List[T]]) -> Observable[T]:
    def flat_map(obj: MutationResponse):
        if obj.errors and len(obj.errors) > 0:
            return rx.throw(obj.errors[0]["message"])
        return rx.from_iterable(mapper(obj.data))

    return pipe(
        ops.map(lambda x: MutationResponse(x)),
        ops.flat_map(flat_map)
    )(observable)

File: gaia-sdk-javascript/src/api/rx.ts
import {MutationResponse, QueryResponse} from "..";
import {Query} from "../graphql/response/type/Query";
import {flatMap, map} from "rxjs/operators";
import {Mutation} from "../graphql/response/type/Mutation";
import {defer, Observable, of, throwError} from "rxjs";

export class Rx {

    static mapQ<T>(observable: Observable<QueryResponse>, mapper: (_: Query) => T): Observable<T> {
        return observable.pipe(
            map(e => JSON.parse(e as string)),
            flatMap(e => {
                if (e.errors && e.errors.length > 0)
                    return throwError(Error(e.errors[0].message));
                return of(mapper(e.data as Query));
            }));
    }

    static flatMapQ<T>(observable: Observable<QueryResponse>, mapper: (_: Query) => [T]): Observable<T> {
        return observable.pipe(
            map(e => JSON.parse(e as string)),
            flatMap(e => {
                if (e.errors && e.errors.length > 0)
                    return throwError(Error(e.errors[0].message));
                return defer(() => mapper(e.data as Query));
            }));
    }

    static mapM<T>(observable: Observable<MutationResponse>, mapper: (_: Mutation) => T): Observable<T> {
        return observable.pipe(
            map(e => JSON.parse(e as string)),
            flatMap(e => {
                if (e.errors && e.errors.length > 0)
                    return throwError(Error(e.errors[0].message));
                return of(mapper(e.data as Mutation));
            }));
    }

    static flatMapM<T>(observable: Observable<MutationResponse>, mapper: (_: Mutation) => [T]): Observable<T> {
        return observable.pipe(
            map(e => JSON.parse(e as string)),
            flatMap(e => {
                if (e.errors && e.errors.length > 0)
                    return throwError(Error(e.errors[0].message));
                return defer(() => mapper(e.data as Mutation));
            }));
    }


}
